programs/stake/src/stake_state.rs

```diff
@@ -595,10 +595,15 @@ impl Stake {
         let rewards = u64::try_from(rewards).unwrap();
 
         // don't bother trying to split if fractional lamports got truncated
-        if rewards == 0 {
-            return None;
-        }
-        let (voter_rewards, staker_rewards, is_split) = vote_state.commission_split(rewards);
+        let (voter_rewards, staker_rewards, is_split) = if rewards == 0 {
+            if fix_stake_deactivate {
```

Comment:
cool. I like this alternative implementation #13233, maybe a lot simpler at the cost of more stake account updates. Yeah, we can use `fix_stake_deactivate` now that we have it.

I'm still thinking about this approach. but, should be good.

Reply:
NGL, I have ulterior motives  :wink:

Reply:
hmm, I think `points == 0` case should be handled like this; otherwise we would allow incorrect rewards for stakes delegating now default then revived vote accounts unless we adjust `calculate_points_and_credits`:

```patch
$ git diff f7218fe3a0a5b237157d02491f8a1f5464c20e4a
diff --git a/programs/stake/src/stake_state.rs b/programs/stake/src/stake_state.rs
index 3cb9a5efe5..061955a033 100644
--- a/programs/stake/src/stake_state.rs
+++ b/programs/stake/src/stake_state.rs
@@ -627,17 +627,24 @@ impl Stake {
             fix_stake_deactivate,
         );
 
-        if points == 0 || point_value.points == 0 {
+        if point_value.points == 0 {
             return None;
         }
+        let rewards = if points == 0 {
+            if fix_stake_deactivate {
+                0
+            } else {
+                return None;
+            }
+        } else {
+            let rewards = points
+                .checked_mul(u128::from(point_value.rewards))
+                .unwrap()
+                .checked_div(point_value.points)
+                .unwrap();
 
-        let rewards = points
-            .checked_mul(u128::from(point_value.rewards))
-            .unwrap()
-            .checked_div(point_value.points)
-            .unwrap();
-
-        let rewards = u64::try_from(rewards).unwrap();
+            u64::try_from(rewards).unwrap()
+        };
 
         // don't bother trying to split if fractional lamports got truncated
         let (voter_rewards, staker_rewards, is_split) = if rewards == 0 {

```

Reply:
I also found this where stake account are missed to be updated: 

```patch
$ git diff 20ac665a83d914f90b4cd2f262cb8d7d8d062f2b
diff --git a/programs/stake/src/stake_state.rs b/programs/stake/src/stake_state.rs
index 061955a033..cf0a7df184 100644
--- a/programs/stake/src/stake_state.rs
+++ b/programs/stake/src/stake_state.rs
@@ -669,7 +669,11 @@ impl Stake {
             // don't collect if we lose a whole lamport somewhere
             //  is_split means there should be tokens on both sides,
             //  uncool to move credits_observed if one side didn't get paid
-            return None;
+            if fix_stake_deactivate {
+                return Some((0, 0, credits_observed));
+            } else {
+                return None;
+            }
         }
 
         Some((staker_rewards, voter_rewards, credits_observed))

```

Reply:
@t-nelson combined with these two findings, I think we are rather better of wrapping this way:

```rust
fn calculate_rewards() {
  let (..., credits_observed) = calculate_points_and_credits(...);
  match _do_calculate_rewards(...) {
     Some(a) => Some(a),
     None => {
        if fix_deactivate_stake {
           (0, 0, credits_reserved)
        } else {
           None
        }
     }
  }
}
```

Reply:
@ryoqun thanks for the careful review!  I'll apply these suggestions locally so I can consider them with more diff context

Reply:
@ryoqun I believe the two proposed changes would re-introduce the issue I pointed out at https://github.com/solana-labs/solana/pull/13233#discussion_r526508417.

To expand, if we update `observed_credits` when `points == 0 && point_value.rewards != 0`, this will deny small stakes delegated to validators with small commission the opportunity to accumulate more points later and eventually be paid out.

Reply:
@t-nelson your understanding is correct. but, I think it's now ok to ignore such small stake accounts for inflation calculation simplicity. 

Otherwise, we'll start to pay normal stakes only for worth a epoch of rewards, on the other hand, we pay very small stakes using any number of past epoch's rewards. I think that's not fair anymore. Also, the calculation will be very complicated. Ideally, I want inflation calculation can be replicated and easily done with rather simple set of formulas on excel. :)

Well, is there any reason i'm missing we should really care for these small stakes?

Reply:
> ... I think that's not fair anymore.

To be really fair in that direction, note that we would need to consider the dynamic set of nano stakes depending on existence of them. That's  because the `PointValue` doesn't consider for unaligned rewards duration for each epoch/stake, under activating/deactivating stakes.

Sounds fun? For me, I'm afraid but it's too complicated. ;)

Reply:
You're :100: about the complexity.  However I see now that there's a missing operational detail which Tyera and I had discussed in DM, but I failed to surface here.  This change should be active for _at least one full epoch_ before inflation is re-enabled.  Doing so ensures that all stake accounts' `observed_credits` are advanced and puts everyone on an even playing field for the first epoch in which inflation rewards are enabled.

As for ignoring small stakes, I'm personally indifferent as to what we do with them.  I was under the impression we had a "no stakes too small" policy, which the stake program seems to historically support.  Perhaps @aeyakovenko can chime in here as to what we've communicated in the past regarding stake size? To reiterate for reference, "small stakes" in this context are those that generate less than one lamport rewards for the validator and/or staker per epoch (1 <= commission <= 99).

Reply:
> However I see now that there's a missing operational detail which Tyera and I had discussed in DM, but I failed to surface here. This change should be active for _at least one full epoch_ before inflation is re-enabled. Doing so ensures that all stake accounts' `observed_credits` are advanced and puts everyone on an even playing field for the first epoch in which inflation rewards are enabled.


@t-nelson Thank for the explanation

well, there are very small and very inactive stake/vote account pairs on mainnet-beta. I too forget to mention this: I've found these unchanging-`credits_observed` stake accounts by running this pr on mb's snapshot with `ledger-tool cap`'s inflation simulation. (For that matter, I've addressing this accounts in #13233, iirc.)

So, we have to do something about other than just waiting enough epochs.

Otherwise, these stake/vote can suddenly be active and start to earn rewards retrospective in unfair way.